src/EventSubscriber/NavigationLock.php

```diff
@@ -0,0 +1,112 @@
+<?php
+
+namespace Drupal\password_enhancements\EventSubscriber;
+
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\Core\Session\SessionManagerInterface;
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\Core\Url;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * Registers a new event subscriber for navigation locking.
+ */
+class NavigationLock implements EventSubscriberInterface {
+
+  use StringTranslationTrait;
+
+  /**
+   * Messenger.
+   *
+   * @var \Drupal\Core\Messenger\MessengerInterface
+   */
+  protected $messenger;
+
+  /**
+   * Module handler.
+   *
+   * @var \Drupal\Core\Extension\ModuleHandlerInterface
+   */
+  protected $moduleHandler;
+
+  /**
+   * Session manager.
+   *
+   * @var \Drupal\Core\Session\SessionManagerInterface
+   */
+  protected $sessionManager;
+
+  /**
+   * Constructs the Legal navigation lock event subscriber.
+   *
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   Messenger.
+   * @param \Drupal\Core\Extension\ModuleHandlerInterface $module_handler
+   *   Module handler.
+   * @param \Drupal\Core\Session\SessionManagerInterface $session_manager
+   *   Session manager.
+   */
+  public function __construct(MessengerInterface $messenger, ModuleHandlerInterface $module_handler, SessionManagerInterface $session_manager) {
+    $this->messenger = $messenger;
+    $this->moduleHandler = $module_handler;
+    $this->sessionManager = $session_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getSubscribedEvents(): array {
+    $events[KernelEvents::REQUEST][] = ['onKernelRequestLockNavigation', 33];
+    return $events;
+  }
+
+  /**
+   * Locks the user to the password change form.
+   *
+   * @param \Symfony\Component\HttpKernel\Event\GetResponseEvent $event
+   *   Response event.
+   */
+  public function onKernelRequestLockNavigation(GetResponseEvent $event): void {
+    /** @var \Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface $attributes_bag */
+    $attributes_bag = $this->sessionManager->getBag('attributes')->getBag();
+    $password_change_required = $attributes_bag->get('password_enhancements_password_change_required');
+
+    if ($password_change_required) {
+      $allowed_paths = [
+        '/user/password-change',
+        '/user/logout',
+      ];
+
+      // Allow altering the allowed paths.
+      $this->moduleHandler->alter('password_enhancements_allowed_paths', $allowed_paths);
```

Comment:
Maybe called it `password_enhancements_password_change_allowed_paths` because it describes in a better way what this alter hook manipulates.

Reply:
You should rather work with routes here as well because as I described above, the path behind a route could be changed. (This is what the `rename_admin_paths` module [does](https://git.drupalcode.org/project/rename_admin_paths/blob/8.x-1.x/src/RenameAdminPathsProcessor.php) in a different way as I described earlier.)

![image](https://user-images.githubusercontent.com/1755573/59485108-be34e980-8e74-11e9-8d92-ffb2af06f155.png)
(Renamed /user* routes to /member*)

This changes the suggested hook name to `password_enhancements_password_change_allowed_routes`.

Reply:
> Maybe called it `password_enhancements_password_change_allowed_paths` because it describes in a better way what this alter hook manipulates.

Then that would be `hook_password_enhancements_password_change_allowed_paths_alter`, so I would rather not add the change string.

Reply:
> You should rather work with routes here as well because as I described above, the path behind a route could be changed. (This is what the `rename_admin_paths` module [does](https://git.drupalcode.org/project/rename_admin_paths/blob/8.x-1.x/src/RenameAdminPathsProcessor.php) in a different way as I described earlier.)
> 
> ![image](https://user-images.githubusercontent.com/1755573/59485108-be34e980-8e74-11e9-8d92-ffb2af06f155.png)
> (Renamed /user* routes to /member*)
> 
> This changes the suggested hook name to `password_enhancements_password_change_allowed_routes`.

I can use the route name, but in the end I have to use their paths as comparison because the current path's route name is not yet available at this point.